pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -1117,3 +1117,348 @@ where
 		}
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use cfg_traits::IdentityCurrencyConversion;
+	use frame_support::{assert_err, assert_ok};
+	use rand::{rngs::StdRng, seq::SliceRandom, SeedableRng};
+
+	use super::*;
+
+	#[derive(Clone, Copy, PartialEq, Debug)]
+	enum CurrencyId {
+		Foreign,
+		Pool,
+	}
+
+	const CONVERSION_RATE: u128 = 3; // 300%
+
+	fn to_pool(foreign_amount: u128) -> u128 {
+		foreign_amount * CONVERSION_RATE
+	}
+
+	fn to_foreign(pool_amount: u128) -> u128 {
+		pool_amount / CONVERSION_RATE
+	}
+
+	struct TestCurrencyConverter;
+
+	impl IdentityCurrencyConversion for TestCurrencyConverter {
+		type Balance = u128;
+		type Currency = CurrencyId;
+		type Error = DispatchError;
+
+		fn stable_to_stable(
+			currency_in: Self::Currency,
+			currency_out: Self::Currency,
+			amount: Self::Balance,
+		) -> Result<Self::Balance, Self::Error> {
+			match (currency_out, currency_in) {
+				(CurrencyId::Foreign, CurrencyId::Pool) => Ok(to_pool(amount)),
+				(CurrencyId::Pool, CurrencyId::Foreign) => Ok(to_foreign(amount)),
+				_ => panic!("Same currency"),
+			}
+		}
+	}
+
+	#[derive(PartialEq)]
+	struct TestConfig;
+
+	impl InvestStateConfig for TestConfig {
+		type Balance = u128;
+		type CurrencyConverter = TestCurrencyConverter;
+		type CurrencyId = CurrencyId;
+	}
+
+	type InvestState = super::InvestState<TestConfig>;
+	type InvestTransition = super::InvestTransition<u128, CurrencyId>;
+
+	#[test]
+	fn increase_with_pool_swap() {
```

Comment:
Could we add a comment documentation to these? They are pretty large and for example `let done_amount` is not used until line 1264 which requires a lot of reading up and down to understand what's going on.

Reply:
It's long because of setting mocks. But actually, it only tests one call. I leave those tests "abandoned" because they overlap with integration tests too much and I feel like I would never end up mocking everything. If we want to continue, I think it needs a refactor to factorize the mock configuration part. So I would refactor instead of documenting it (but in case we chose to go in that path)